plugins/de.cau.cs.kieler.klighd.lsp/src/de/cau/cs/kieler/klighd/lsp/utils/KGraphElementIDGenerator.xtend

```diff
@@ -0,0 +1,269 @@
+     */
+    def String getId(KGraphElement element) {
+        if (element === null) {
+            return ""
+        }
+        var String id = null
+        
+        // if the ID was already calculated, use that
+        if (elementToIdMap.get(element) !== null) {
+            return elementToIdMap.get(element)
+        }
+        
+        // the root node is just called $root
+        val parent = element.eContainer as KGraphElement
+        var String parentId = null
+        if (parent !== null) {
+            parentId = getId(parent)
+        } else {
+            return ID_SEPARATOR + 'root'
+        }
+        
+        // use a prefix depending on the class of the element + the {@link KIdentifier} as id if an identifier is
+        // defined, otherwise make up a new id based on the position in the model hierarchy with a Separator not
+        // appearing in the {@link KIdentifier} (the ID Separator).
+        var String elementId = null
+        
+        val identifier = element.data.filter(KIdentifier)
+        var char elementSeparator
+        
+        switch (element) {
+            KNode: {
+                elementSeparator = NODE_SEPARATOR
+            }
+            KEdge: {
+                elementSeparator = EDGE_SEPARATOR
+            }
+            KLabel: {
+                elementSeparator = LABEL_SEPARATOR
+            }
+            KPort: {
+                elementSeparator = PORT_SEPARATOR
+            }
+            default: {
+                throw new IllegalArgumentException("Can not generate an id for element of type " + element.class)
+            }
+        }
+        
+        if (identifier.empty) {
+            elementId = "" + ID_SEPARATOR + elementSeparator + element.hashCode
+        } else {
+            elementId = elementSeparator + identifier.head.id
+        }
+        
+        val baseId = parentId + ID_SEPARATOR + elementId
+        // If the KIdentifier is not unique between its siblings, make the ID unique with a counter in the end.
+        id = baseId
+        for (var cnt = 2; idToElementMap.containsKey(id); cnt++) {
+            id = baseId + ID_SEPARATOR + ID_SEPARATOR + "copy" + cnt
+        }
+        elementToIdMap.put(element, id)
+        idToElementMap.put(id, element)
+        return id
+    }
+    
+    /**
+     * Utility method to find a {@link KGraphElement} from a map between KGraphElements and {@link SModelElement}s by
+     * the ID stored in the target SModelElement.
+     * 
+     * @param k2sMap The map between KGraphElements and SModelElements.
+     * @param id The unique ID of one SModelElement in the map.
+     * @return The KGraphElement that is the key in the given map with a matching ID in its value.
+     */
+    static def findElementById(Map<KGraphElement, SModelElement> k2sMap, String id) {
+        // TODO: maybe store the map also the other way, so this can be done more efficiently.
+        // Or have a map from IDs to their originating KGraphElements.
```

Comment:
Do we still want to do this?

Reply:
Yes. The code below is just inefficient as it walks through all elements. This did not have priority yet, as it is only ever used once and is not performance critical.

Reply:
I suggest to implement it or to add a ticket for this.